towhee/dataframe/dataframe.py

```diff
@@ -35,19 +37,171 @@ def __init__(self, name: str, data: List[Tuple[Variable]] = None):
                 `Dataframe`. These tuples can be interpreted as being direct outputs
                 into downstream operators.
         """
-        # self._iter = ScalarIterator(df)
         self._name = name
-        self._data = data
+        self._data = data if data is not None else []
+        self._registered_iter = []
+        self._start_index = 0
+        self._total = 0
+        self._end = False
+
+        self._lock = threading.Lock()
+        self._registered_lock = threading.Lock()
```

Comment:
What is the purpose of this registered lock?

Reply:
A dataframe's iter will register in multithread, so I use `_registered_lock` to make thread-safe